## Importing the Library and the Dataset

In [1]:
import pandas as pd # For reading the data and manipulating it
import numpy as np # For mathematical computations.

import matplotlib.pyplot as plt # For plotting 
import seaborn as sns # For plotting 

In [2]:
cars = pd.read_excel('Data_Train.xlsx') # reading the training file into cars dataframe

cars.sample(5) # viewing 5 random samples of the data

Name    Location  Year  Kilometers_Driven  \
4902          Maruti Baleno Alpha CVT      Mumbai  2018               3000   
25       Honda City Corporate Edition      Mumbai  2012              51920   
922                Maruti Celerio VXI   Bangalore  2015              74000   
1825  Maruti Vitara Brezza VDi Option  Coimbatore  2017              71256   
4683     Hyundai Verna 1.6 CRDi AT SX       Kochi  2016              42876   

     Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
4902    Petrol    Automatic      First   21.4 kmpl  1197 CC   83.1 bhp    5.0   
25      Petrol       Manual      First   16.8 kmpl  1497 CC  116.3 bhp    5.0   
922     Petrol       Manual      First   23.1 kmpl   998 CC  67.04 bhp    5.0   
1825    Diesel       Manual      First   24.3 kmpl  1248 CC   88.5 bhp    5.0   
4683    Diesel    Automatic      First  19.08 kmpl  1582 CC  126.2 bhp    5.0   

      New_Price  Price  
4902  10.3 Lakh   8.85  
25          NaN   4.25  
922         NaN   4.15  
1825        NaN   9.58  
4683        NaN   9.31

## Prepairing the Dataset for Analysis

#### Info about the `cars` dataframe

In [3]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6019 entries, 0 to 6018
Data columns (total 13 columns):
Name                 6019 non-null object
Location             6019 non-null object
Year                 6019 non-null int64
Kilometers_Driven    6019 non-null int64
Fuel_Type            6019 non-null object
Transmission         6019 non-null object
Owner_Type           6019 non-null object
Mileage              6017 non-null object
Engine               5983 non-null object
Power                5983 non-null object
Seats                5977 non-null float64
New_Price            824 non-null object
Price                6019 non-null float64
dtypes: float64(2), int64(2), object(9)
memory usage: 611.4+ KB


* Our `cars` data has **6019** rows and **13** columns. And right now it seems like :
* __Mileage__ has **2** missing values.
* __Engine__ & __Power__ have  **36** missing values
* __Seats has__ **47** missing values.
* __New_Price__ has **5195** missing values, which means that **87%** rows have its value missing. And hence we will drop it later. 

#### Name

In [4]:
print('No. of unique enteries : ', len(cars.Name.unique()))

No. of unique enteries :  1876


The name of the car follows the pattern -> [Brand Name] [Car Name]    
Also I believe that cars from certain brands will have higher resale value than other brands. 
So, I am creating a new col `Brand` from the `Name` col.

In [5]:
cars['Brand'] = [name.split(' ')[0].capitalize() for name in cars.Name] # Extracting Brand Name from car name.

cars.sample(3)

Name   Location  Year  Kilometers_Driven  \
199   Renault Duster 85PS Diesel RxL      Delhi  2014              90000   
5239                 Audi Q5 2.0 TDI  Hyderabad  2013              72000   
5819       Renault KWID RXT Optional      Kochi  2017              21954   

     Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
199     Diesel       Manual      First  19.87 kmpl  1461 CC   83.8 bhp    5.0   
5239    Diesel    Automatic      First  14.16 kmpl  1968 CC  174.3 bhp    5.0   
5819    Petrol       Manual      First  25.17 kmpl   799 CC   53.3 bhp    5.0   

      New_Price  Price    Brand  
199         NaN   5.35  Renault  
5239        NaN  29.00     Audi  
5819  4.78 Lakh   3.70  Renault

#### Brand

Available brands of cars in the dataset

In [6]:
print(cars.Brand.value_counts().index.values) # List of brands available

['Maruti' 'Hyundai' 'Honda' 'Toyota' 'Mercedes-benz' 'Volkswagen' 'Ford'
 'Mahindra' 'Bmw' 'Audi' 'Tata' 'Skoda' 'Renault' 'Chevrolet' 'Nissan'
 'Land' 'Jaguar' 'Fiat' 'Mitsubishi' 'Mini' 'Volvo' 'Porsche' 'Jeep'
 'Datsun' 'Isuzu' 'Force' 'Smart' 'Bentley' 'Ambassador' 'Lamborghini']


Replacing `Land` by `Land Rover`

In [7]:
cars.Brand.replace('Land', 'Land Rover', inplace = True) # Correcting a spelling error

print(cars.Brand.value_counts().index.values)

['Maruti' 'Hyundai' 'Honda' 'Toyota' 'Mercedes-benz' 'Volkswagen' 'Ford'
 'Mahindra' 'Bmw' 'Audi' 'Tata' 'Skoda' 'Renault' 'Chevrolet' 'Nissan'
 'Land Rover' 'Jaguar' 'Fiat' 'Mitsubishi' 'Mini' 'Volvo' 'Porsche' 'Jeep'
 'Datsun' 'Isuzu' 'Force' 'Bentley' 'Smart' 'Ambassador' 'Lamborghini']


#### Location

In [8]:
print(cars.Location.value_counts().index.values) # List of cities

['Mumbai' 'Hyderabad' 'Kochi' 'Coimbatore' 'Pune' 'Delhi' 'Kolkata'
 'Chennai' 'Jaipur' 'Bangalore' 'Ahmedabad']


All good here !

#### Year

In [9]:
print(cars.Year.value_counts().index.values) # List of years

[2014 2015 2016 2013 2017 2012 2011 2010 2018 2009 2008 2007 2019 2006
 2005 2004 2003 2002 2001 1998 2000 1999]


The oldest car is of year **1999**  and the latest is from **2019**, hence cars from a range of 20 years.

#### Kilometers_Driven

I will first rename the header name from **Kilometers_Driven** to **Km** for easiness.

In [10]:
cars.rename(columns = {         
    'Kilometers_Driven' : 'Km',
}, inplace = True)

In [11]:
cars.Km.describe()

count    6.019000e+03
mean     5.873838e+04
std      9.126884e+04
min      1.710000e+02
25%      3.400000e+04
50%      5.300000e+04
75%      7.300000e+04
max      6.500000e+06
Name: Km, dtype: float64

#### Fuel_Type

I will first rename the header name from **Fuel_Type** to **Fuel** for easiness.

In [12]:
cars.rename(columns = {
    'Fuel_Type' : 'Fuel',
}, inplace = True)

In [13]:
cars.Fuel.value_counts()

Diesel      3205
Petrol      2746
CNG           56
LPG           10
Electric       2
Name: Fuel, dtype: int64

#### Transmission

In [14]:
cars.Transmission.value_counts()  # Viewing no. of auto cars and no. of manual cars.

Manual       4299
Automatic    1720
Name: Transmission, dtype: int64

Nothing to change here !

#### Owner_Type

In [15]:
cars.rename(columns = {
    'Owner_Type' : 'Owner'
}, inplace = True)

In [16]:
cars.Owner.value_counts()

First             4929
Second             968
Third              113
Fourth & Above       9
Name: Owner, dtype: int64

In [17]:
cars.sample(3)

Name    Location  Year      Km  \
1217                           Honda City ZX GXi      Mumbai  2005   56000   
3546     Mahindra Scorpio VLX 4WD AIRBAG AT BSIV  Coimbatore  2011  108905   
5998  Volkswagen Polo IPL II 1.2 Petrol Highline        Pune  2010   70290   

        Fuel Transmission   Owner     Mileage   Engine     Power  Seats  \
1217  Petrol       Manual  Second   17.7 kmpl  1497 CC    78 bhp    5.0   
3546  Diesel    Automatic  Second  11.79 kmpl  2179 CC   120 bhp    8.0   
5998  Petrol       Manual  Second  17.24 kmpl  1198 CC  73.9 bhp    5.0   

     New_Price  Price       Brand  
1217       NaN   1.25       Honda  
3546       NaN   6.21    Mahindra  
5998       NaN   2.96  Volkswagen

#### Mileage

In [18]:
cars[cars.Fuel == 'LPG'][:2] # Viewing cars with fuel type LPG

Name   Location  Year     Km Fuel  \
5    Hyundai EON LPG Era Plus Option  Hyderabad  2012  75000  LPG   
936      Maruti Wagon R LXI LPG BSIV  Hyderabad  2012  72000  LPG   

    Transmission  Owner     Mileage  Engine     Power  Seats New_Price  Price  \
5         Manual  First  21.1 km/kg  814 CC  55.2 bhp    5.0       NaN   2.35   
936       Manual  First  26.2 km/kg  998 CC  58.2 bhp    5.0       NaN   2.85   

       Brand  
5    Hyundai  
936   Maruti

In [19]:
cars[cars.Fuel == 'CNG'][:2] # Viewing cars with fuel type CNG

Name Location  Year     Km Fuel Transmission  Owner  \
0    Maruti Wagon R LXI CNG   Mumbai  2010  72000  CNG       Manual  First   
127  Maruti Wagon R LXI CNG     Pune  2013  89900  CNG       Manual  First   

        Mileage  Engine      Power  Seats New_Price  Price   Brand  
0    26.6 km/kg  998 CC  58.16 bhp    5.0       NaN   1.75  Maruti  
127  26.6 km/kg  998 CC  58.16 bhp    5.0       NaN   3.25  Maruti

In [20]:
cars[cars.Fuel == 'Electric'] # Viewing cars with fuel type Electric

Name Location  Year     Km      Fuel Transmission  \
4446       Mahindra E Verito D4  Chennai  2016  50000  Electric    Automatic   
4904  Toyota Prius 2009-2016 Z4   Mumbai  2011  44000  Electric    Automatic   

      Owner Mileage   Engine   Power  Seats   New_Price  Price     Brand  
4446  First     NaN    72 CC  41 bhp    5.0  13.58 Lakh  13.00  Mahindra  
4904  First     NaN  1798 CC  73 bhp    5.0         NaN  12.75    Toyota

In [21]:
cars[cars.Fuel == 'Diesel'][:2] # Viewing cars with fuel type Diesel

Name Location  Year     Km    Fuel  \
1  Hyundai Creta 1.6 CRDi SX Option     Pune  2015  41000  Diesel   
3                 Maruti Ertiga VDI  Chennai  2012  87000  Diesel   

  Transmission  Owner     Mileage   Engine      Power  Seats New_Price  Price  \
1       Manual  First  19.67 kmpl  1582 CC  126.2 bhp    5.0       NaN   12.5   
3       Manual  First  20.77 kmpl  1248 CC  88.76 bhp    7.0       NaN    6.0   

     Brand  
1  Hyundai  
3   Maruti

In [22]:
cars[cars.Fuel == 'Petrol'][:2] # Viewing cars with fuel type Petrol

Name Location  Year     Km    Fuel Transmission  Owner  \
2       Honda Jazz V  Chennai  2011  46000  Petrol       Manual  First   
10  Maruti Ciaz Zeta    Kochi  2018  25692  Petrol       Manual  First   

       Mileage   Engine       Power  Seats   New_Price  Price   Brand  
2    18.2 kmpl  1199 CC    88.7 bhp    5.0   8.61 Lakh   4.50   Honda  
10  21.56 kmpl  1462 CC  103.25 bhp    5.0  10.65 Lakh   9.95  Maruti

##### Correcting the units of Mileage and converting dtype to floats.

In [23]:
def convert_unit(car):
    if (car.Fuel == 'Diesel' or car.Fuel == 'Petrol'):
        mileage = car.Mileage.strip(' kmpl')
    elif car.Fuel == 'CNG':
        mileage = car.Mileage.strip(' km/kg')
    elif car.Fuel == 'LPG':
        mileage = car.Mileage.strip(' km/kg')
    elif car.Fuel == 'Electric':
        if car.Brand == 'Mahindra':
            mileage = 110
        else :
            mileage = 24
    mileage = float(mileage)
    return mileage

In [24]:
cars['Mileage_Converted'] = cars.apply(convert_unit, axis = 1)

cars.drop(columns = 'Mileage', inplace = True)
cars.rename(columns = {'Mileage_Converted' : 'Mileage'}, inplace = True)

cars.sample(5)

Name   Location  Year  \
1385                                Honda City 1.5 GXI       Pune  2004   
4401               Tata Tiago 1.2 Revotron XZ WO Alloy  Hyderabad  2017   
5646                         Maruti Ciaz ZDi Plus SHVS  Ahmedabad  2015   
3297                             Nissan Micra XL Primo     Jaipur  2010   
5525  Mercedes-Benz New C-Class C 220 CDI BE Avantgare  Bangalore  2016   

          Km    Fuel Transmission   Owner   Engine       Power  Seats  \
1385  115000  Petrol       Manual  Second      NaN         NaN    NaN   
4401   16500  Petrol       Manual   First  1199 CC      84 bhp    5.0   
5646   52000  Diesel       Manual   First  1248 CC    88.5 bhp    5.0   
3297   75055  Petrol       Manual   First  1198 CC      75 bhp    5.0   
5525   37944  Diesel    Automatic   First  2143 CC  167.62 bhp    5.0   

      New_Price  Price          Brand  Mileage  
1385        NaN   1.50          Honda     0.00  
4401  6.29 Lakh   5.00           Tata    23.84  
5646        NaN   6.90         Maruti    28.09  
3297        NaN   2.25         Nissan    18.06  
5525        NaN  35.00  Mercedes-benz    14.84

In [25]:
cars.Mileage.describe()

count    6019.000000
mean       18.151198
std         4.732671
min         0.000000
25%        15.170000
50%        18.160000
75%        21.100000
max       110.000000
Name: Mileage, dtype: float64

In [26]:
print("No. of cars with mileage 0 (zero) :", len(cars[cars.Mileage == 0]))

No. of cars with mileage 0 (zero) : 68


I guess, these actually represent null values. So I will replace them with NaN.

In [27]:
cars.Mileage.replace(0, np.nan, inplace = True)

#### Engine

In [28]:
cars[cars.Engine.isnull()]

Name    Location  Year      Km  \
194                             Honda City 1.5 GXI   Ahmedabad  2007   60006   
208                           Maruti Swift 1.3 VXi     Kolkata  2010   42001   
733                           Maruti Swift 1.3 VXi     Chennai  2006   97800   
749                   Land Rover Range Rover 3.0 D      Mumbai  2008   55001   
1294                             Honda City 1.3 DX       Delhi  2009   55005   
1327                          Maruti Swift 1.3 ZXI   Hyderabad  2015   50295   
1385                            Honda City 1.5 GXI        Pune  2004  115000   
1460  Land Rover Range Rover Sport 2005 2012 Sport  Coimbatore  2008   69078   
2074                          Maruti Swift 1.3 LXI        Pune  2011   24255   
2096                     Hyundai Santro LP zipPlus  Coimbatore  2004   52146   
2264                           Toyota Etios Liva V        Pune  2012   24500   
2325                      Maruti Swift 1.3 VXI ABS        Pune  2015   67000   
2335                          Maruti Swift 1.3 VXi      Mumbai  2007   55000   
2530                       BMW 5 Series 520d Sedan       Kochi  2014   64158   
2542               Hyundai Santro GLS II - Euro II   Bangalore  2011   65000   
2623                       BMW 5 Series 520d Sedan        Pune  2012   95000   
2668                          Maruti Swift 1.3 VXi     Kolkata  2014   32986   
2737                             Maruti Wagon R Vx      Jaipur  2001  200000   
2780               Hyundai Santro GLS II - Euro II        Pune  2009  100000   
2842               Hyundai Santro GLS II - Euro II   Bangalore  2012   43000   
3272                       BMW 5 Series 520d Sedan      Mumbai  2008   81000   
3404                          Maruti Swift 1.3 VXi      Jaipur  2006  125000   
3520                       BMW 5 Series 520d Sedan       Delhi  2012   90000   
3522               Hyundai Santro GLS II - Euro II       Kochi  2012   66400   
3810                   Honda CR-V AT With Sun Roof     Kolkata  2013   27000   
4011                        Fiat Punto 1.3 Emotion        Pune  2011   45271   
4152                  Land Rover Range Rover 3.0 D      Mumbai  2003   75000   
4229                        Hyundai Santro Xing XG   Bangalore  2005   79000   
4577                       BMW 5 Series 520d Sedan       Delhi  2012   72000   
4604                     Honda Jazz Select Edition        Pune  2011   98000   
4697                        Fiat Punto 1.2 Dynamic       Kochi  2017   17941   
4712                        Hyundai Santro Xing XG        Pune  2003   80000   
4952                        Fiat Punto 1.4 Emotion     Kolkata  2010   47000   
5015                          Maruti Swift 1.3 VXi       Delhi  2006   63000   
5185                          Maruti Swift 1.3 LXI       Delhi  2012   52000   
5270                            Honda City 1.5 GXI   Bangalore  2002   53000   

        Fuel Transmission           Owner Engine Power  Seats New_Price  \
194   Petrol       Manual           First    NaN   NaN    NaN       NaN   
208   Petrol       Manual           First    NaN   NaN    NaN       NaN   
733   Petrol       Manual           Third    NaN   NaN    NaN       NaN   
749   Diesel    Automatic          Second    NaN   NaN    NaN       NaN   
1294  Petrol       Manual           First    NaN   NaN    NaN       NaN   
1327  Petrol       Manual           First    NaN   NaN    NaN       NaN   
1385  Petrol       Manual          Second    NaN   NaN    NaN       NaN   
1460  Petrol       Manual           First    NaN   NaN    NaN       NaN   
2074  Petrol       Manual           First    NaN   NaN    NaN       NaN   
2096  Petrol       Manual           First    NaN   NaN    NaN       NaN   
2264  Petrol       Manual          Second    NaN   NaN    NaN       NaN   
2325  Petrol       Manual           First    NaN   NaN    NaN       NaN   
2335  Petrol       Manual          Second    NaN   NaN    NaN       NaN   
2530  Diesel    Automatic          

##### The cars with NaN as value in Engine col have missing values for Power, Seats, New Price and some have missing values for Mileage as well, so I am removing these cars from the dataset

In [29]:
cars_to_drop_index = cars[cars.Engine.isnull()].index.values # indexes of cars with missing engine vals.

In [30]:
cars.drop(cars_to_drop_index, inplace = True) 

In [31]:
def remove_CC(car):
    engine = car.Engine
    if (engine[-2:] == 'CC') :
        engine = engine.strip('CC')
    engine = int(engine)
    return engine

In [32]:
cars['Engine_'] = cars.apply(remove_CC, axis = 1)

cars.drop(columns = 'Engine', inplace = True)
cars.rename(columns = {'Engine_' : 'Engine'}, inplace = True)

cars.sample(3)

Name   Location  Year      Km    Fuel  \
4621      Tata Manza Aura Quadrajet BS IV  Hyderabad  2011   70259  Diesel   
154   Skoda Superb Elegance 2.0 TDI CR AT       Pune  2012  136997  Diesel   
5128          Honda Jazz 1.2 V CVT i VTEC      Kochi  2016   30534  Petrol   

     Transmission  Owner      Power  Seats New_Price  Price  Brand  Mileage  \
4621       Manual  First   88.8 bhp    5.0       NaN   2.95   Tata    21.12   
154     Automatic  First  138.1 bhp    5.0       NaN   8.50  Skoda    17.20   
5128    Automatic  First   88.7 bhp    5.0       NaN   7.42  Honda    19.00   

      Engine  
4621    1248  
154     1968  
5128    1199

In [33]:
cars.Engine.describe()

count    5983.000000
mean     1621.276450
std       601.355233
min        72.000000
25%      1198.000000
50%      1493.000000
75%      1984.000000
max      5998.000000
Name: Engine, dtype: float64

#### Power

In [34]:
cars.Power.describe()

count       5983
unique       372
top       74 bhp
freq         235
Name: Power, dtype: object

In [35]:
print('No. of null or missing values :', cars.Power.isnull().sum())

No. of null or missing values : 0


In [36]:
len(cars[(cars.Power == 'null bhp')])

107

##### Some of the cars have `null` bhp as power, I will replace that with np.nan and from rest will remove `bhp`.

In [37]:
def remove_bhp(car):
    bhp = car.Power
    if (bhp == 'null bhp') :
        bhp = np.nan
    elif (bhp[-3:] == 'bhp'):
        bhp = bhp.strip(' bhp')
        bhp = float(bhp)
    return bhp

In [38]:
cars['Power_'] = cars.apply(remove_bhp, axis = 1)

cars.drop(columns = 'Power', inplace = True)
cars.rename(columns = {'Power_' : 'Power'}, inplace = True)

cars.sample(3)

Name   Location  Year     Km    Fuel  \
4236  Audi A4 35 TDI Premium Plus      Kochi  2015  13506  Diesel   
481          Ford Figo Diesel EXI    Chennai  2011  75000  Diesel   
444            Hyundai EON D Lite  Hyderabad  2014  20600  Petrol   

     Transmission   Owner  Seats  New_Price  Price    Brand  Mileage  Engine  \
4236    Automatic   First    5.0  50.4 Lakh  21.66     Audi    18.25    1968   
481        Manual  Second    5.0        NaN   2.25     Ford    20.00    1399   
444        Manual   First    5.0        NaN   2.60  Hyundai    21.10     814   

       Power  
4236  187.74  
481    68.00  
444    55.20

In [39]:
print('No. of null or missing values :', cars.Power.isnull().sum())

No. of null or missing values : 107


#### Seats

In [40]:
cars.Seats.describe()

count    5977.000000
mean        5.278735
std         0.808840
min         0.000000
25%         5.000000
50%         5.000000
75%         5.000000
max        10.000000
Name: Seats, dtype: float64

Min seats are 0 (zero)! 

In [41]:
cars[cars.Seats == 0]

Name   Location  Year      Km    Fuel  \
3999  Audi A4 3.2 FSI Tiptronic Quattro  Hyderabad  2012  125000  Petrol   

     Transmission  Owner  Seats New_Price  Price Brand  Mileage  Engine  Power  
3999    Automatic  First    0.0       NaN   18.0  Audi     10.5    3197    NaN

Since there is only one car with 0 seats, I will correct its data from internet.

In [42]:
cars.loc[cars.Seats == 0, ['Seats']] = 5

In [43]:
cars.Seats.value_counts()

5.0     5015
7.0      674
8.0      134
4.0       99
6.0       31
2.0       16
10.0       5
9.0        3
Name: Seats, dtype: int64

#### New_Price

As mentioned earlier, the `New_Price` col has about `87%` rows with missing values, So I will remove it from the dataset.

In [44]:
cars.drop(columns = ['New_Price'], inplace = True)

In [45]:
cars.describe().transpose()

count          mean           std      min      25%       50%  \
Year     5983.0   2013.383085      3.249102  1998.00   2011.0   2014.00   
Km       5983.0  58684.183186  91503.344783   171.00  33965.5  53000.00   
Seats    5977.0      5.279572      0.805960     2.00      5.0      5.00   
Price    5983.0      9.496263     11.200462     0.44      3.5      5.65   
Mileage  5926.0     18.366564      4.346963     6.40     15.3     18.20   
Engine   5983.0   1621.276450    601.355233    72.00   1198.0   1493.00   
Power    5876.0    113.253050     53.874957    34.20     75.0     97.70   

              75%        max  
Year      2016.00     2019.0  
Km       73000.00  6500000.0  
Seats        5.00       10.0  
Price        9.95      160.0  
Mileage     21.10      110.0  
Engine    1984.00     5998.0  
Power      138.10      560.0

##### End of prepairing data for analysis. 
saving the data to `pre-analysis-data.csv`.

In [46]:
cars.to_csv('pre-analysis-data.csv', index = False)